sdk/core/core-rest-pipeline/src/util/helpers.ts

```diff
@@ -1,22 +1,71 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
+import { AbortError, AbortSignalLike } from "@azure/abort-controller";
+
 /**
  * A constant that indicates whether the environment the code is running is Node.JS.
  * @internal
  */
 export const isNode =
   typeof process !== "undefined" && Boolean(process.version) && Boolean(process.versions?.node);
 
+const StandardAbortMessage = "The operation was aborted.";
+
 /**
- * A wrapper for setTimeout that resolves a promise after t milliseconds.
- * @internal
- * @param t - The number of milliseconds to be delayed.
+ * A wrapper for setTimeout that resolves a promise after delayInMs milliseconds.
+ * @param delayInMs - The number of milliseconds to be delayed.
  * @param value - The value to be resolved with after a timeout of t milliseconds.
- * @returns Resolved promise
+ * @param options - The options for delay - currently abort options
+ *   @param abortSignal - The abortSignal associated with containing operation.
+ *   @param abortErrorMsg - The abort error message associated with containing operation.
+ * @returns - Resolved promise
  */
-export function delay<T>(t: number, value?: T): Promise<T | void> {
-  return new Promise((resolve) => setTimeout(() => resolve(value), t));
+export function delay<T>(
```

Comment:
I don’t know if the delay function is the best place to put this code.
However, I understand why it would make sense to put it in the delay code.
So I think it’s ok!

Do we have other examples of the delay function handling the abort signal?
I wonder if this is going to be a pattern moving forward.

Reply:
Should this `delay` function with abortSignal support be in core-rest-pipelines or perhaps in another core package?

Reply:
I have not written this from scratch, this was from `core-http`.

And we also have a similar(identical) method in `core-amqp` and `service-bus` also relies on it I think.

Reply:
> Should this delay function with abortSignal support be in core-rest-pipelines or perhaps in another core package?

It is now added in core-rest-pipeline.
I would prefer `core-util`, but it is not a dependency of `core-rest-pipeline`.

Reply:
Anyway, I like the design! A lot! Just food for thought.

Reply:
ohh I understand now! Alright